src/Mod/Part/App/TopoShapeExpansion.cpp

```diff
@@ -3161,6 +3358,244 @@ TopoShape& TopoShape::makeElementRefine(const TopoShape& shape, const char* op,
+
+                if (!bspline2.IsNull()) {
+                    gp_Trsf transf = heloc2.Transformation();
+                    bspline2->Transform(transf);  // apply original transformation to control points
+                    // Store Underlying Geometry
+                    curves.push_back(bspline2);
+                }
+                else {
+                    // BRepBuilderAPI_NurbsConvert failed, try ShapeConstruct_Curve now
+                    ShapeConstruct_Curve scc;
+                    Handle(Geom_BSplineCurve) spline =
+                        scc.ConvertToBSpline(c_geom, u1, u2, Precision::Confusion());
+                    if (spline.IsNull()) {
+                        Standard_Failure::Raise(
+                            "A curve was not a B-spline and could not be converted into one.");
+                    }
+                    gp_Trsf transf = heloc2.Transformation();
+                    spline->Transform(transf);  // apply original transformation to control points
+                    curves.push_back(spline);
+                }
+            }
+        }
+
+        GeomFill_BSplineCurves aSurfBuilder;  // Create Surface Builder
+
+        if (edges.size() == 2) {
+            aSurfBuilder.Init(curves[0], curves[1], fstyle);
+        }
+        else if (edges.size() == 3) {
+            aSurfBuilder.Init(curves[0], curves[1], curves[2], fstyle);
+        }
+        else if (edges.size() == 4) {
+            aSurfBuilder.Init(curves[0], curves[1], curves[2], curves[3], fstyle);
+        }
+
+        aSurface = aSurfBuilder.Surface();
+    }
+
+    BRepBuilderAPI_MakeFace aFaceBuilder;
+    Standard_Real v1, v2;
+    // transfer surface bounds to face
+    aSurface->Bounds(u1, u2, v1, v2);
+
+    aFaceBuilder.Init(aSurface, u1, u2, v1, v2, Precision::Confusion());
+
+    TopoShape aFace(0, Hasher, aFaceBuilder.Face());
+
+    if (!aFaceBuilder.IsDone()) {
+        FC_THROWM(Base::CADKernelError, "Face unable to be constructed");
+    }
+    if (aFace.isNull()) {
+        FC_THROWM(Base::CADKernelError, "Resulting Face is null");
+    }
+
+    // TODO:  Is this correct?  makeElementBSplineFace is new ( there is no corresponding non element
+    // version of this operation.  It appears to be reasonable for the BRepBuilderAPI_MakeFace to
+    // return more edges than we sent in, and in that case we should be resetting all the new edge
+    // elementmaps and then mapping the sub elements.  It is potentially counterintuitive to the user
+    // in that edges can be created "from nowhere", which may be why this effectively disallows them.
+    // But it does so with segfaults.
+    auto newEdges = aFace.getSubTopoShapes(TopAbs_EDGE);
+    if (newEdges.size() != edges.size()) {
+        FC_WARN("Face edge count mismatch");
+    }
+    else {
+        int ind = 0;
+        for (auto& edge : newEdges) {
+            edge.resetElementMap(edges[ind++].elementMap());
+        }
```

Comment:
Should we also be resetting the element maps of the edges in the case of a mismatch (the edges we actually have, that is)?

Reply:
I like that approach, just wanted more eyeballs on this condition before attempting it.  Thus the comment. 
 Thanks!